### PR TITLE
Remove unwrap()s from context.rs

### PR DESCRIPTION
**Pull request by alphastrata**
I removed all the ones I could.

```cargo test --doc``` 97 Passes, 2 Ignored, 0 Failing.

Hope it's acceptable.

### CODE REVIEW

**Comment by bheisler** [comment]
Thanks for the pull request! It might be a while before I can review this properly.

File: src/context.rs [hunk 1]
```diff
@@ -37,70 +37,80 @@
...
+//!   let context =
+//!     Context::create_and_push(ContextFlags::MAP_HOST | ContextFlags::SCHED_AUTO, device)?;
+//!   let mut join_handles = vec![];
 //!
-//! let mut join_handles = vec![];
-//! for _ in 0..4 {
+//!   for _ in 0..4 {
 //!     let unowned = context.get_unowned();
 //!     let join_handle = std::thread::spawn(move || {
-//!         CurrentContext::set_current(&unowned).unwrap();
```
Comment (bheisler) [inline src/context.rs]: Did you mean to change the indentation here?

File: src/context.rs [hunk 2]
```diff
@@ -37,70 +37,80 @@
...
+//! for device in Device::devices()? {
+//!     let device = device?;
+//!     let ctx =
+//!         Context::create_and_push(ContextFlags::MAP_HOST | ContextFlags::SCHED_AUTO, device)?;
+//!     ContextStack::pop()?;
+//!     contexts.push(ctx);
+//! }
+//! CurrentContext::set_current(&contexts[0])?;
+//!
+//!  // Call Rustacuda functions which will use the context
```
[the remark this answers is not in the document]
Comment (bheisler) [reply src/context.rs]: Also, it looks like the indentation of this line was changed. Was that intended?

File: src/context.rs [hunk 3]
```diff
@@ -689,14 +749,18 @@ impl CurrentContext {
...
-    /// # let device = Device::get_device(0).unwrap();
-    /// let context = Context::create_and_push(ContextFlags::MAP_HOST | ContextFlags::SCHED_AUTO, device).unwrap();
-    /// CurrentContext::set_resource_limit(ResourceLimit::StackSize, 2048).unwrap();
+    /// # use std::error::Error;
+    /// # 
+    /// # fn main () -> Result<(), Box<dyn Error>>{
+    /// # rustacuda::init(rustacuda::CudaFlags::empty())?;
+    /// # let device = Device::get_device(0)?;
+    /// let context = Context::create_and_push(ContextFlags::MAP_HOST | ContextFlags::SCHED_AUTO, device)?;
+    /// let priority_range = CurrentContext::get_stream_priority_range()?;
```
Comment (bheisler) [inline src/context.rs]: Another stray call to get_stream_priority_range here.

**Comment by bheisler** [comment]
For the most part this looks good, but there are a bunch of minor things that will need to be cleaned up.

- There are a number of what appear to be copy-paste errors and other minor things.
- There is also some trailing whitespace in the examples which is causing the Rustfmt check in the CI builds to fail.
- In other circumstances, it would make sense to add another level of indentation when wrapping a function around some code, but since the reader of the examples can't see the `fn main` code, for these examples the indentation shouldn't be changed.

Other comments:

- I won't be too picky about it, but in future it would be a good idea to remove commits like your test and test2 commits from the history.
- It's OK that you've only updated the Context examples, but I would recommend mentioning that in the PR comment or the issue so that people know what the pull request does or does not change without having to read the code.

File: src/context.rs [hunk 4]
```diff
@@ -37,70 +37,80 @@
 //! For most commmon uses (one device, one OS thread) it should suffice to create a single context:
 //!
 //! ```
-//! use rustacuda;
 //! use rustacuda::device::Device;
 //! use rustacuda::context::{Context, ContextFlags};
+//! use std::error::Error;
```
[the remark this answers is not in the document]
Comment (alphastrata) [reply src/context.rs]: Understood.

File: src/context.rs (hunk 1 above)
Comment (alphastrata) [reply src/context.rs]: Nope, my mistake, I assumed `cargo fmt` was cleaning up my code... apparently not?

File: src/context.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (alphastrata) [reply src/context.rs]: I think I just lifted it from your comment as that one was a little hard for me.

https://github.com/bheisler/RustaCUDA/issues/1#issuecomment-447573452

File: src/context.rs [hunk 5]
```diff
@@ -236,13 +246,17 @@ impl Context {
...
-    /// let context = Context::create_and_push(ContextFlags::MAP_HOST | ContextFlags::SCHED_AUTO, device).unwrap();
+    /// # use rustacuda::device::Device;
+    /// # use rustacuda::context::{Context, ContextFlags};
+    /// # use std::error::Error;
+    /// #
+    /// # fn main () -> Result<(), Box<dyn Error>>{
+    ///    rustacuda::init(rustacuda::CudaFlags::empty())?;
+    ///    let device = Device::get_device(0)?;
+    ///    let context = Context::create_and_push(ContextFlags::MAP_HOST | ContextFlags::SCHED_AUTO, device)?;
+    ///    let version = context.get_api_version()?;
```
[the remark this answers is not in the document]
Comment (alphastrata) [reply src/context.rs]: Nope, just silly copy+paste errors as you suspected.

[the remark this answers is not in the document]
Comment (alphastrata) [reply src/context.rs]: I will remove the spacing to make it inline with the others too.

File: src/context.rs [hunk 6]
```diff
@@ -648,14 +704,18 @@ impl CurrentContext {
...
-    /// # let device = Device::get_device(0).unwrap();
-    /// let context = Context::create_and_push(ContextFlags::MAP_HOST | ContextFlags::SCHED_AUTO, device).unwrap();
-    /// CurrentContext::set_cache_config(CacheConfig::PreferL1).unwrap();
+    /// # use std::error::Error;
+    /// # 
+    /// # fn main () -> Result<(), Box<dyn Error>>{
+    /// # rustacuda::init(rustacuda::CudaFlags::empty())?;
+    /// # let device = Device::get_device(0)?;
+    /// let context = Context::create_and_push(ContextFlags::MAP_HOST | ContextFlags::SCHED_AUTO, device)?;
+    /// let priority_range = CurrentContext::get_stream_priority_range()?;
```
[the remark this answers is not in the document]
Comment (alphastrata) [reply src/context.rs]: Nope.

**Comment by alphastrata** [comment]
Should I just refork your repo & then submit a new pull request, once I've made those cleanups and possibly others?

**Comment by alphastrata** [comment]
Whoops, while messing around with git I accidentally closed this.
Sorry.

**Comment by alphastrata** [comment]
This should be good to go.

If it's acceptable, I'll start working on the other `unwrap()` calls in `device.rs` etc ~

**Comment by rusch95** [comment]
I split up device.rs, so make sure to pull from master before doing that

**Comment by bheisler** [comment]
Thanks for the pull request!

**Pull request merged**